NRPT-214: New Record Types

dhlevi opened the pull request:
Adding new record types for NRPTI:

- Annual Report
- Dam Safety Inspection
- Report
- Certificate Amendment
- Correspondence 

includes API, common, and UI components.

All record types contain the mineGuid ref, but currently there's no way to set this via the screens and it is not part of the existing ticket spec. We'll have to include a method of associating the guid for manual creation of these and other record types for use in BCMI.

There's a lot of code to look over and I'm sure I've missed a few spots or made mistakes, so I'm pinging everyone to have a look!

See tickets [NRPT-214](https://bcmines.atlassian.net/browse/NRPT-214) (the UI) and [NRPT-284](https://bcmines.atlassian.net/browse/NRPT-284) (The API)

KitArmstrong commented on angular/projects/admin-nrpti/src/app/records/annual-reports/annual-report-add-edit/annual-report-add-edit.component.ts:
```diff
@@ -0,0 +1,354 @@
+      this.lastEditedSubText = `Added on ${this.utils.convertJSDateToString(new Date(this.currentRecord.dateAdded))}`;
+    }
+    for (const flavour of this.currentRecord.flavours) {
+      switch (flavour._schemaName) {
```
Nit: maybe just an `if` in this case?

KitArmstrong commented on angular/projects/admin-nrpti/src/app/records/annual-reports/annual-report-add-edit/annual-report-add-edit.component.ts:
```diff
@@ -0,0 +1,354 @@
+  }
+
+  navigateToDetails() {
+    this.router.navigate(['records', 'annual-reports', this.currentRecord._id, 'detail']);
```
Without the `/` in front of records this will navigate relative the route it is called from. Is that intended? Depending on where this screen is it might work as expected without the `/` too.

dhlevi replied:
For the routers here I just copied it exactly from the other details/add-edit screens. Happy to make a change, but if so I should make the change to all of them, not just the five new ones... they all seem to work as expected for now though!

KitArmstrong commented on angular/projects/admin-nrpti/src/app/records/annual-reports/annual-report-add-edit/annual-report-add-edit.component.ts:
```diff
@@ -0,0 +1,354 @@
+
+    if (!this.isEditing) {
+      this.factoryService.createAnnualReport(annualReport).subscribe(async res => {
+        console.log(res);
```
Log.

KitArmstrong commented on angular/projects/admin-nrpti/src/app/records/annual-reports/annual-report-add-edit/annual-report-add-edit.component.ts:
```diff
@@ -0,0 +1,354 @@
+          this.factoryService
+        );
+
+        console.log(docResponse);
```
Log.

KitArmstrong commented on angular/projects/admin-nrpti/src/app/records/annual-reports/annual-report-bcmi-detail/annual-report-bcmi-detail.component.ts:
```diff
@@ -0,0 +1,114 @@
+      .pipe(
+        takeUntil(this.ngUnsubscribe),
+        catchError(error => {
+          console.log('Publish error:', error);
```
Log.

KitArmstrong commented on angular/projects/admin-nrpti/src/app/records/certificate-amendments/certificate-amendments-add-edit/certificate-amendments-add-edit.component.ts:
```diff
@@ -0,0 +1,414 @@
+
+    if (!this.isEditing) {
+      this.factoryService.createCertificateAmendment(certificateAmendment).subscribe(async res => {
+        console.log(res);
```
Log.

KitArmstrong commented on angular/projects/admin-nrpti/src/app/records/certificate-amendments/certificate-amendments-add-edit/certificate-amendments-add-edit.component.ts:
```diff
@@ -0,0 +1,414 @@
+          this.factoryService
+        );
+
+        console.log(docResponse);
```
Log.

KitArmstrong commented on angular/projects/admin-nrpti/src/app/records/certificate-amendments/certificate-amendments-add-edit/certificate-amendments-add-edit.component.ts:
```diff
@@ -0,0 +1,414 @@
+          this.factoryService
+        );
+
+        console.log(docResponse);
```
Log.

KitArmstrong commented on angular/projects/admin-nrpti/src/app/records/correspondences/correspondence-add-edit/correspondence-add-edit.component.ts:
```diff
@@ -0,0 +1,412 @@
+
+    if (!this.isEditing) {
+      this.factoryService.createCorrespondence(correspondence).subscribe(async res => {
+        console.log(res);
```
Log

dhlevi commented:
> Tested well locally. Nice to see everything following the same approach. I left a couple comments about logs and errors. I did not note them in every component, but might need to check them all.

Fixed the log statements and other notes above. Checked all of the new pages and updated where appropriate across the board.

Early on in development, I thought it would be a much nicer idea to convert a lot of the duplicate flavour stuff, record screens and functions scattered in various other components (the huge case statements, ex.) into more generic common components. It's starting to get fairly unruly in some spots and much of it would be easy enough to convert to single functions with a type. Not the time to do it with this ticket, but it might be worth a technical debt ticket to see about updating a lot of this stuff into shared/common functions and components.

mtCarto commented on angular/projects/admin-nrpti/src/app/services/factory.service.ts:
```diff
@@ -418,6 +422,96 @@ export class FactoryService {
+    const outboundObject = {
+      correspondences: [correspondence]
+    };
+    console.log(outboundObject);
```
remove

mtCarto commented on angular/projects/admin-nrpti/src/app/services/factory.service.ts:
```diff
@@ -418,6 +422,96 @@ export class FactoryService {
+    const outboundObject = {
+      annualReports: [report]
+    };
+    console.log(outboundObject);
```
remove

mtCarto commented on angular/projects/admin-nrpti/src/app/services/factory.service.ts:
```diff
@@ -418,6 +422,96 @@ export class FactoryService {
+    const outboundObject = {
+      damSafetyInspections: [damSafetyInspection]
+    };
+    console.log(outboundObject);
```
remove

dhlevi commented:
Extra console logging removed
